# Прикладная линейная алгебра

## Домашняя контрольная работа 1

## Поздняков Виталий, Вариант 18

___

#### 1. Найдите сингулярное разложение для

$$A =
\begin{pmatrix}
9 & 5 & 0 \\
−3 & −7 & −6 \\
3 & 7 & 6 \\
−3 & −7 & −6 \\
\end{pmatrix}$$

#### Решение

$$A_{4\times 3} = Q_{4 \times 4}\Sigma_{4 \times 3} R_{3 \times 3}^*$$

In [1]:
import numpy as np

In [2]:
A = np.array([[ 9,  5,  0],
              [-3, -7, -6],
              [ 3,  7,  6],
              [-3, -7, -6]])
A.T @ A

array([[108, 108,  54],
       [108, 172, 126],
       [ 54, 126, 108]])

$$A^*A =
\begin{pmatrix}
108 & 108 & 54 \\
108 & 172 & 126 \\
54 & 126 & 108
\end{pmatrix}
$$

$$|A^*A - \lambda E| = 
\begin{vmatrix}
108 - \lambda & 108 & 54 \\
108 & 172 - \lambda & 126 \\
54 & 126 & 108 - \lambda
\end{vmatrix} =
-\lambda^3 + 388 \lambda^2 - 18360 \lambda = 0
$$

$$\begin{matrix}
\lambda_1 = 0 & \sigma_3 = \sqrt 0 = 0 \\
\lambda_2 = 194 - 2\sqrt{4819} = 55.16 & \sigma_2 = \sqrt{194 - 2\sqrt{4819}} = 7.43 \\
\lambda_3 = 194 + 2\sqrt{4819} = 332.84 & \sigma_1 = \sqrt{194 + 2\sqrt{4819}} = 18.24 \\
\end{matrix}$$

$$\sum = 
\begin{pmatrix}
\sigma_1 & 0 & 0 \\
0 & \sigma_2 & 0 \\
0 & 0 & \sigma_3 \\
0 & 0 & 0
\end{pmatrix} =
\begin{pmatrix}
18.24 & 0 & 0 \\
0 & 7.43 & 0 \\
0 & 0 & 0 \\
0 & 0 & 0
\end{pmatrix}$$


In [3]:
lambda2 = 194 - 2 * (4819) ** (1/2)
sigma2 = lambda2 ** (1/2)
lambda3 = 194 + 2 * (4819) ** (1/2)
sigma1 = lambda3 ** (1/2)
Sigma = np.zeros([4, 3])
Sigma[0, 0] = sigma1
Sigma[1, 1] = sigma2

Найдем собственный вектор $x_1$ для $\sigma_1$ ($\lambda_3 = 332.84$)

In [4]:
# приведение матрицы к ступенчатому виду методом прямого хода Гаусса
def echelon(matrix):
    matrix = matrix.astype(float)
    shape = matrix.shape[0]
    for i in range(shape - 1):
        for j in range(i + 1, shape):
            matrix[j] = (
                matrix[j] - matrix[i] * matrix[j, i] / matrix[i, i])
    for i in range(shape - 1):
        if matrix[i, i] != 0:
            matrix[i] = matrix[i] * (1 / matrix[i, i])
    return matrix

# нормирование вектора
def normed(vector):
    norm = 0
    for x in vector:
        norm += x * x
    return vector / norm ** (1/2)

In [5]:
(A.T @ A - np.eye(3) * lambda3).round(2)

array([[-224.84,  108.  ,   54.  ],
       [ 108.  , -160.84,  126.  ],
       [  54.  ,  126.  , -224.84]])

In [6]:
eq1 = echelon(A.T @ A - np.eye(3) * lambda3)
eq1.round(2)

array([[ 1.  , -0.48, -0.24],
       [-0.  ,  1.  , -1.39],
       [ 0.  ,  0.  ,  0.  ]])

$$\begin{cases}
\nu_1 - 0.48 \nu_2 - 0.24 \nu_3 = 0 \\
\nu_2 - 1.39 \nu_3 = 0 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = 0.48 \nu_2 + 0.24 \nu_3 \\
\nu_2 = 1.39 \nu_3 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = 0.48 \cdot 1.39 \nu_3 + 0.24 \nu_3 \\
\nu_2 = 1.39 \nu_3 \\
\nu_3 = \nu_3
\end{cases}$$

$$
x_1 = \begin{pmatrix}
0.91 \\
1.39 \\
1
\end{pmatrix}$$

Нормируем вектор $x_1$

In [7]:
x1 = np.array([-eq1[0, 1] * -eq1[1, 2] + -eq1[0, 2], -eq1[1, 2], 1])
x1 = normed(x1)
x1.round(2)

array([0.47, 0.72, 0.51])

$$ x_1  =\begin{pmatrix}
0.47 \\
0.72 \\
0.52
\end{pmatrix}$$

Найдем собственный вектор $x_2$ для $\sigma_2$ ($\lambda_2 = 55.16$)

In [8]:
(A.T @ A - np.eye(3) * lambda2).round(2)

array([[ 52.84, 108.  ,  54.  ],
       [108.  , 116.84, 126.  ],
       [ 54.  , 126.  ,  52.84]])

In [9]:
eq2 = echelon(A.T @ A - np.eye(3) * lambda2)
eq2.round(2)

array([[ 1.  ,  2.04,  1.02],
       [-0.  ,  1.  , -0.15],
       [ 0.  ,  0.  ,  0.  ]])

$$\begin{cases}
\nu_1 + 2.04 \nu_2 + 1.02 \nu_3 = 0 \\
\nu_2 - 0.15 \nu_3 = 0 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = - 2.04 \nu_2 - 1.02 \nu_3 \\
\nu_2 = 0.15 \nu_3 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = - 2.04 \cdot 0.15 \nu_3 - 1.02 \nu_3 \\
\nu_2 = 0.15 \nu_3 \\
\nu_3 = \nu_3
\end{cases}$$

$$
x_2 = \begin{pmatrix}
-1.326 \\
0.15 \\
1
\end{pmatrix}$$

Нормируем вектор $x_2$

In [10]:
x2 = np.array([-eq2[0, 1] * -eq2[1, 2] + -eq2[0, 2], -eq2[1, 2], 1])
x2 = normed(np.array([-1.326, 0.15, 1]))
x2.round(2)

array([-0.8 ,  0.09,  0.6 ])

$$ x_2  =\begin{pmatrix}
-0.8 \\
0.09 \\
0.6
\end{pmatrix}$$

Найдем собственный вектор $x_3$ для $\sigma_3$ ($\lambda_1 = 0$)

In [11]:
A.T @ A

array([[108, 108,  54],
       [108, 172, 126],
       [ 54, 126, 108]])

In [12]:
eq3 = echelon(A.T @ A)
eq3

array([[1.   , 1.   , 0.5  ],
       [0.   , 1.   , 1.125],
       [0.   , 0.   , 0.   ]])

$$\begin{cases}
\nu_1 + \nu_2 + 0.5 \nu_3 = 0 \\
\nu_2 + 1.125 \nu_3 = 0 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = - \nu_2 - 0.5 \nu_3 \\
\nu_2 = -1.125 \nu_3 \\
\nu_3 = \nu_3
\end{cases} \Rightarrow
\begin{cases}
\nu_1 = 1.125 \nu_3 - 0.5 \nu_3 \\
\nu_2 = -1.125 \nu_3 \\
\nu_3 = \nu_3
\end{cases}$$

$$
x_3 = \begin{pmatrix}
0.625 \\
-1.125 \\
1
\end{pmatrix}$$

Нормируем вектор $x_3$

In [13]:
x3 = np.array([-eq3[0, 1] * -eq3[1, 2] + -eq3[0, 2], -eq3[1, 2], 1])
x3 = normed(x3)
x3.round(2)

array([ 0.38, -0.69,  0.61])

$$ x_3  =\begin{pmatrix}
0.38 \\
-0.69 \\
0.61
\end{pmatrix}$$

$$R = 
\begin{pmatrix}
0.47 & -0.8 & 0.38 \\ 
0.72 & 0.09 & -0.69 \\ 
0.52 & 0.6 & 0.61
\end{pmatrix}$$

Найдем матрицу $Q$

In [14]:
q1 = A @ x1 / sigma1
q1.round(2)

array([ 0.43, -0.52,  0.52, -0.52])

$$
q_1 = \frac{A  x_1 }{\sigma_1} =
\begin{pmatrix}
0.43 \\
-0.52 \\
0.52 \\
-0.52
\end{pmatrix}$$

In [15]:
q2 = A @ x2 / sigma2
q2.round(2)

array([-0.9 , -0.25,  0.25, -0.25])

$$
q_2 = \frac{A  x_2 }{\sigma_1} =
\begin{pmatrix}
-0.9 \\
-0.25 \\
0.25 \\
-0.25
\end{pmatrix}$$

Для нахождения $q_3$ и $q_4$ применим алгоритм Грама-Шмидта. Добавим линейно независимые векторы, чтобы получить произвольный базис. Пусть $q'_3 = (0, 0, 0, 1)$ и $q'_4 = (0, 0, 1, 0)$

In [16]:
q_3 = np.array([0, 0, 0, 1])
q_4 = np.array([0, 0, 1, 0])
q3 = normed(q_3 - q1 @ q_3 * q1 - q2 @ q_3 * q2)
q4 = normed(q_4 - q1 @ q_4 * q1 - q2 @ q_4 * q2 - q3 @ q_4 * q3)
Q = np.array([q1, q2, q3, q4]).T
Q.round(2)

array([[ 0.43, -0.9 , -0.  ,  0.  ],
       [-0.52, -0.25, -0.41,  0.71],
       [ 0.52,  0.25,  0.41,  0.71],
       [-0.52, -0.25,  0.82,  0.  ]])

Ответ: $$Q\Sigma R^* = 
\begin{pmatrix}
0.43 & -0.9 & 0 & 0 \\
-0.52 & -0.25 & -0.41 & 0.71 \\
0.52 & 0.25 & 0.41 & 0.71 \\
-0.52 & -0.25 & 0.82 & 0 \\
\end{pmatrix}
\begin{pmatrix}
18.24 & 0 & 0 \\
0 & 7.43 & 0 \\
0 & 0 & 0 \\
0 & 0 & 0
\end{pmatrix}
\begin{pmatrix}
0.47 & 0.72 & 0.52 \\
-0.8 & 0.09 & 0.6 \\
0.38 & -0.69 & 0.61
\end{pmatrix}$$

Проверим

In [17]:
np.set_printoptions(suppress=True)
Q @ Sigma @ np.array([x1, x2, x3])

array([[ 8.99032942,  5.00105181, -0.00282735],
       [-2.9954151 , -7.00053018, -6.00622383],
       [ 2.9954151 ,  7.00053018,  6.00622383],
       [-2.9954151 , -7.00053018, -6.00622383]])

Ответ верный с точностью до 1 знака после запятой

#### 2. Найдите какое-нибудь разложение полного ранга и псевдообратную матрицу для

$$A = \begin{pmatrix}
1 & 2 & 2 \\
3 & 8 & 5 \\
-1 & -4 & -1 \\
5 & 14 & 8
\end{pmatrix}$$

#### Решение

Найдем $B$ и $C$ такие, чтобы $A = BC$

Приведем матрицу $A$ к каноническому виду методом Гаусса

$$\begin{pmatrix}
1 & 2 & 2 \\
3 & 8 & 5 \\
-1 & -4 & -1 \\
5 & 14 & 8
\end{pmatrix} =
\begin{pmatrix}
1 & 0 & 3 \\
0 & 1 & -0.5
\end{pmatrix}
= C$$

$rkA=2$, значит первые 2 столбца из матрицы $A$ — это матрица $B$

$$B = \begin{pmatrix}
1 & 2 \\
3 & 8 \\
-1 & -4 \\
5 & 14
\end{pmatrix}$$

Проверим

In [18]:
B = np.array([[1, 2], 
              [3, 8], 
              [-1, -4], 
              [5, 14]])
C = np.array([[1, 0, 3], 
              [0, 1, -0.5]])
B @ C

array([[ 1.,  2.,  2.],
       [ 3.,  8.,  5.],
       [-1., -4., -1.],
       [ 5., 14.,  8.]])

Ответ верный

Найдем псевдообратную матрицу $A^+=C^+B^+$

Найдем $B^+$. $rkB = 2$ полный столбцовый ранг, значит

In [19]:
B_plus = np.linalg.inv(B.T @ B) @ B.T
B_plus

array([[ 1.  ,  0.5 ,  1.5 , -0.  ],
       [-0.35, -0.15, -0.55,  0.05]])

$$B^+ = (B^*B)^{-1}B^* = 
\begin{pmatrix}
1 & 0.5 & 1.5 & 0 \\
-0.35 & -0.15 & -0.55 & 0.05
\end{pmatrix}
$$

Найдем $C^+$. $rkC = 2$ полный строчный ранг, значит

In [20]:
C_plus = C.T @ np.linalg.inv(C @ C.T)
C_plus.round(2)

array([[ 0.12,  0.15],
       [ 0.15,  0.98],
       [ 0.29, -0.05]])

$$C^+ = C^*(CC^*)^{-1} = 
\begin{pmatrix}
0.12 & 0.15 \\
0.15 & 0.98 \\
0.29 & -0.05
\end{pmatrix}
$$

In [21]:
A_plus = C_plus @ B_plus
A_plus.round(2)

array([[ 0.07,  0.04,  0.1 ,  0.01],
       [-0.2 , -0.07, -0.32,  0.05],
       [ 0.31,  0.15,  0.47, -0.  ]])

Ответ: $$A^+=C^+B^+=
\begin{pmatrix}
 0.07 &  0.04 &  0.1  &  0.01 \\
 -0.2 & -0.07 & -0.32 &  0.05 \\
 0.31 &  0.15 &  0.47 &  0  \\
\end{pmatrix}
$$

Проверим

In [22]:
A = np.array([[1, 2, 2], 
              [3, 8, 5], 
              [-1, -4, -1], 
              [5, 14, 8]])
A @ A_plus @ A

array([[ 1.,  2.,  2.],
       [ 3.,  8.,  5.],
       [-1., -4., -1.],
       [ 5., 14.,  8.]])

Ответ верный

#### 3. Среди всех приближений решения следующей системы по методу наименьших квадратов найдите вектор наименьшей длины.

$$\begin{cases}
4 · x + 14 · y + 2 · z − 6 · t = 4 \\
3 · x + 8 · y + 5 · z + 1 · t = 1 \\
2 · x + 1 · y + 5 · z + 5 · t = 3 \\
2 · x + 2 · y + 8 · z + 8 · t = 9
\end{cases}$$

#### Решение

В матричном виде: $AX=B$, где

$$A = 
\begin{pmatrix}
4 & 14 & 2 & -6 \\
3 & 8 & 5 & 1 \\
2 & 1 & 5 & 5 \\
2 & 2 & 8 & 8 \\
\end{pmatrix}; 
X = 
\begin{pmatrix}
x \\
y \\
z \\
t
\end{pmatrix};
B = 
\begin{pmatrix}
4 \\
1 \\
3 \\
9
\end{pmatrix}$$

Решение будем искать в виде $\tilde{X}=A^{+}B$, где $A^+ = (A^*A)^{-1}A^*$

In [23]:
A = np.array([[4, 14, 2, -6], 
              [3, 8, 5, 1], 
              [2, 1, 5, 5],
              [2, 2, 8, 8]])
A_plus = np.linalg.inv(A.T @ A) @ A.T
A_plus.round(2)

array([[ 0.25, -0.06,  1.4 , -0.38],
       [-0.01,  0.03, -0.34,  0.09],
       [-0.06,  0.06, -0.28,  0.19],
       [-0.02,  0.  ,  0.03,  0.03]])

In [24]:
B = np.array([4, 1, 3, 9])
(A_plus @ B).round(2)

array([ 1.78, -0.2 ,  0.66,  0.28])

Ответ: $$\begin{cases}
x = 1.78 \\
y = -0.2 \\
z = 0.66 \\
t = 0.28
\end{cases}$$

#### 4. Для многочлена $−2x^3 + 3x^2 − 5x − 5$ найдите наилучшее равномерное приближение многочленом второй степени на отрезке $[ -2 , 4 ]$.

#### Решение

По теореме Чебышева

$$ \Vert f(x) - P_{n-1}(x) \Vert_\infty \rightarrow \min_{x \in [a, b]} $$ 
при 
$$f(x) - P_{n-1}(x) = \bar T_n(x) \Rightarrow P_{n-1}(x) = f(x) - \bar T_n(x)$$ 
где
$$\bar T_n(x) = \frac{\alpha(b-a)^n}{2^{2n-1}}T_n\left(\frac{2x-(b+a)}{b-a}\right)$$
и $\alpha$ — старший коэффициент в многочлене $f(x)$

Для нашей задачи имеем

$$f(x) = −2x^3 + 3x^2 − 5x − 5$$
$$n = 3$$
$$a = -2, b = 4$$
$$\alpha = -2$$

Тогда

$$\bar T_3(x) = -2\frac{(4+2)^3}{2^{2\cdot3-1}}T_3\left(\frac{2x-(4-2)}{4+2}\right) = -2\frac{6^3}{2^{5}}T_3\left(\frac{2x-2}{6}\right)$$

Многочлен Чебышева 3 степени:

$$T_3(x)=4x^3-3x \Rightarrow T_3\left(\frac{2x-2}{6}\right)=4\left(\frac{2x-2}{6}\right)^3-3\left(\frac{2x-2}{6}\right)$$
$$T_3\left(\frac{2x-2}{6}\right)=4\left(\frac{8x^3-24x^2+24x-8}{6^3}\right)-3\left(\frac{2x-2}{6}\right)$$
$$T_3\left(\frac{2x-2}{6}\right)=\frac{8x^3-24x^2+24x-8}{54}-\frac{2x-2}{2}$$
$$T_3\left(\frac{2x-2}{6}\right)=\frac{8x^3-24x^2+24x-8-54x+54}{54}$$
$$T_3\left(\frac{2x-2}{6}\right)=\frac{8x^3-24x^2-30x+46}{54}$$

$$\bar T_3(x) = -2 \frac{6^3}{2^5}\cdot\frac{8x^3-24x^2-30x+46}{54} = -2 \frac{8x^3-24x^2-30x+46}{8} $$
$$\bar T_3(x) = -2\left(x^3 - 3x^2 - \frac{15}{4}x+\frac{23}{4}\right) = -2x^3 + 6x^2 + \frac{15}{2}x - \frac{23}{2}$$
$$P_2(x) = −2x^3 + 3x^2 − 5x − 5 + 2x^3 - 6x^2 - \frac{15}{2}x + \frac{23}{2}$$

Ответ:
$$P_2(x) = - 3x^2 - \frac{25}{2}x + \frac{13}{2}$$

<img src='2019-11-04_23-38-20.png' align='center' width=500>

#### 5. При каком $q$ уравнение $−2qxz+x^2 +xy(1−2q)+y^2(4q+1)−yz+ z^2 (4q + 1) = 1$ задает единичную окружность относительно какой-то нормы? Найдите норму вектора $(1, 1, 1)$ в зависимости от $q$.

#### Решение

Построим матрицу квадратичной формы

$$M =
\begin{pmatrix}
1 & \frac{1-2q}{2} & -q \\
\frac{1-2q}{2} & 4q+1 & -\frac{1}{2} \\
-q & -\frac{1}{2} & 4q + 1
\end{pmatrix}
$$

Найдем значения $q$, при которых квадратичная форма положительно определена

$\vartriangle_1 = 1 > 0$ при любых $q$

$\vartriangle_2 = -x^2+5x+\frac{3}{4} > 0$ при $q \in [-0.15, 5.14]$ (приближенно)

$\vartriangle_3 = -8x^3+17x^2+\frac{17}{2}x+\frac{1}{2} > 0$ при $q \in [-\infty, -0.36] \cup [-0.07, 2.55]$ (приближенно)

Таким образом, квадратичная форма положительно определена и уравнение задает единичную окружность относительно нормы при $q \in [-0.07, 2.55]$

Рассмотрим отрезок $OA$ c точками $O = (0, 0, 0)$ и $A = (1, 1, 1)$ на плоскости $z=ax+by+c$ 

$$\begin{cases}
0 = a \cdot 0 + b \cdot 0 + c \\
1 = a \cdot 1 + b \cdot 1 + c
\end{cases} \Rightarrow
\begin{cases}
c = 0 \\
a + b = 1
\end{cases}$$

Рассмотрим два случая $(a=1\land b=0)$ и $(a=0\land b=1)$. Подставим полученные коэффициенты в уравнение плоскости, тогда прямая будет определяться плоскостями $z = y$ и $z = x$. Найдем точку пересечения этих плоскостей и эллипсоида: 

$$\begin{cases}
z = x \\
z = y \\
−2qxz+x^2 +xy(1−2q)+y^2(4q+1)−yz+ z^2 (4q + 1) = 1
\end{cases}$$

$$−2qz^2+z^2 +z^2(1−2q)+z^2(4q+1)−z^2+ z^2 (4q + 1) = 1$$
$$z^2(−2q+1 + 1−2q + 4q+1 − 1 + 4q + 1) = 1$$
$$z^2(3 + 4q) = 1$$
$$z^2 = \frac{1}{3 + 4q}$$
$$z = x = y = \pm \sqrt \frac{1}{3 + 4q}$$

Точка пересечения с $OA$ должна иметь координаты $>0$, значит

$$ B = \left( \sqrt \frac{1}{3 + 4q}, \sqrt \frac{1}{3 + 4q}, \sqrt \frac{1}{3 + 4q} \right)$$

Пусть $v = (1, 1, 1)$, тогда норма вектора $v$:

$$\nu(v) = \frac{|OA|}{|OB|} = \frac{\sqrt 3}{\sqrt \frac{3}{3+4q}} = \sqrt{3+4q}$$

Ответ: 

Уравнение задает единичную окружность относительно нормы при $q \in [-0.07, 2.55]$ (приближенно). Норма вектора $(1, 1, 1)$ равняется $\sqrt{3+4q}$.